# Udacity Data Science - Project 2 - Data Wrangling WeRateDogs

### Activities
1. Gather
2. Assess
3. Clean
4. Explore
5. Conclusion

## 1. Gather

In [159]:
import tweepy
import pandas as pd
import numpy as np
import requests
import json
import time

### Auth Tweepy

In [160]:
consumer_key = 'hY02aNzLmpOgJbKFDzQZosVCq'
consumer_secret = '0TO8DI7VdJFJDl56pUXDTZotIJYPZWBHid4eLFoPfyKCNQ3IX4'
access_token = '173806049-c7csKJxD7uvzPWsJFX9Q0FgtEhTYpPrbbbgxzz2d'
access_secret = 'bzwwhXrIWcZGAliD3UUFhWKC8i4HYPfIHObQrTnKy1pwt'
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)
api = tweepy.API(auth,  wait_on_rate_limit=True, wait_on_rate_limit_notify=True)
# iniciar a API

### Checking connectivity

In [161]:
api

### Importing Files

In [162]:
df_tweets = pd.read_csv('twitter-archive-enhanced.csv')
df_dog_image_predictions = pd.read_csv('image-predictions.tsv', sep='\t')

## Assess

### Tweets

In [163]:
df_tweets.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

In [164]:
# Visualizando a estrutura de um tweet da API
tweet = api.get_status(666020888022790149, tweet_mode='extended')
tweet._json

{'created_at': 'Sun Nov 15 22:32:08 +0000 2015',
 'id': 666020888022790149,
 'id_str': '666020888022790149',
 'full_text': 'Here we have a Japanese Irish Setter. Lost eye in Vietnam (?). Big fan of relaxing on stair. 8/10 would pet https://t.co/BLDqew2Ijj',
 'truncated': False,
 'display_text_range': [0, 131],
 'entities': {'hashtags': [],
  'symbols': [],
  'user_mentions': [],
  'urls': [],
  'media': [{'id': 666020881337073664,
    'id_str': '666020881337073664',
    'indices': [108, 131],
    'media_url': 'http://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg',
    'media_url_https': 'https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg',
    'url': 'https://t.co/BLDqew2Ijj',
    'display_url': 'pic.twitter.com/BLDqew2Ijj',
    'expanded_url': 'https://twitter.com/dog_rates/status/666020888022790149/photo/1',
    'type': 'photo',
    'sizes': {'medium': {'w': 960, 'h': 720, 'resize': 'fit'},
     'thumb': {'w': 150, 'h': 150, 'resize': 'crop'},
     'large': {'w': 960, 'h': 720, 'resize': 'fit

In [10]:
# Adicionando todos os ID's em uma variável
tweet_ids = df_tweets.tweet_id

# Lista com os ids que não puderam ser recuperados.
tweet_erro_ids = []

# Lista com os tweets retornados pela.
tweet_from_api = []  

# iniciar contador para controlar o tempo entre as chamadas na API do Twitter.
start = time.time()

# iniciar loop entre os ids
for id in tweet_ids:
    try:
        # recupera a informação de um Tweet pelo ID.
        tweet = api.get_status(id)
    
        # Armazena o retorno do tweet.
        tweet_from_api.append({  
            'tweet_id': str(id),
            'favorite_count': tweet.favorite_count,
            'retweet_count': tweet.retweet_count
        })
    except tweepy.TweepError as e:
        print(e)
        # Em caso de erro armazenar para tratamento posterior.
        tweet_erro_ids.append(id)
            
# finalizar contador
end = time.time()
    
# mensagens de conclusão
print('Concluído! O processo levou: ' + str(end - start) + ' segundos')

[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status fou

Rate limit reached. Sleeping for: 605


[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]


Rate limit reached. Sleeping for: 620


Concluído! O processo levou: 1983.4147996902466 segundos


In [224]:
tweet_from_api[0]

{'tweet_id': '892420643555336193',
 'favorite_count': 37589,
 'retweet_count': 8198}

In [225]:
# Armazenando os dados da API em um arquivo local, tweet_json.txt.
with open('tweet_json.txt', 'w') as file:  
    json.dump(tweet_from_api, file)

In [15]:
# Id's que não puderam ser recuperados na API.
tweet_erro_ids

[888202515573088257,
 873697596434513921,
 872668790621863937,
 872261713294495745,
 869988702071779329,
 866816280283807744,
 861769973181624320,
 856602993587888130,
 845459076796616705,
 844704788403113984,
 842892208864923648,
 837012587749474308,
 827228250799742977,
 812747805718642688,
 802247111496568832,
 775096608509886464,
 770743923962707968,
 754011816964026368,
 680055455951884288]

In [165]:
# importando o dados do arquivo local com o retorno da API
df_tweets_api = pd.read_json('tweet_json.txt')

In [166]:
# Visualizando leitura do arquivo
df_tweets_api.head()

favorite_count  retweet_count            tweet_id
0           37589           8198  892420643555336192
1           32309           6060  892177421306343424
2           24348           4009  891815181378084864
3           40933           8357  891689557279858688
4           39126           9049  891327558926688256

## 2. Assess

### 2.1 - Data

Foram disponibilizados três fontes de dados para este projeto.
1. Um arquivo com Id's e informações de tweets, fornecido e preparado diretamente pela Udacity
2. Um arquivo com dados de predições da classificação de cachorros em imagens.
3. Recuperação de mais informações de Tweets através da fonte 1 diretamente na API do Twitter.

# 2.1.1 - First Datasource - df_tweets

In [167]:
df_tweets.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [168]:
# Visualização dos tipos de dados contidos no DataFrame
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [212]:
# Quantidade de linhas e colunas
df_tweets.shape

(2356, 17)

In [170]:
# Atributos com valores unicos.
df_tweets.nunique()

tweet_id                      2356
in_reply_to_status_id           77
in_reply_to_user_id             31
timestamp                     2356
source                           4
text                          2356
retweeted_status_id            181
retweeted_status_user_id        25
retweeted_status_timestamp     181
expanded_urls                 2218
rating_numerator                40
rating_denominator              18
name                           957
doggo                            2
floofer                          2
pupper                           2
puppo                            2
dtype: int64

In [215]:
df_tweets.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

Existem apenas 4 tipos de source, porém o dado está como HTMl.

In [220]:
df_tweets.text.value_counts()

13/10 for breakdancing puppo @shibbnbot                                                                                                                        1
RT @dog_rates: "Yep... just as I suspected. You're not flossing." 12/10 and 11/10 for the pup not flossing https://t.co/SuXcI9B7pQ                             1
This is Rupert. You betrayed him with bath time but he forgives you. Cuddly af 13/10 https://t.co/IEARC2sRzC                                                   1
This is Kirby. He's a Beneblip Cumberpat. Pretty heckin rare. 11/10 would put my face against his face https://t.co/fd6uucghY6                                 1
This is Cassie. She goes door to door trying to find the owner of this baguette. No luck so far. 10/10 https://t.co/e8bj97CisO                                 1
Say hello to Blakely. He thinks that's a hat. Silly pupper. That's a nanner. 9/10 https://t.co/UwOV1jqKRt                                                      1
This is Loomis. He's the leader of

A coluna text, possui textos e links.

In [171]:
# Acesssando coluna expanded_urls
df_tweets.expanded_urls

0       https://twitter.com/dog_rates/status/892420643...
1       https://twitter.com/dog_rates/status/892177421...
2       https://twitter.com/dog_rates/status/891815181...
3       https://twitter.com/dog_rates/status/891689557...
4       https://twitter.com/dog_rates/status/891327558...
5       https://twitter.com/dog_rates/status/891087950...
6       https://gofundme.com/ydvmve-surgery-for-jax,ht...
7       https://twitter.com/dog_rates/status/890729181...
8       https://twitter.com/dog_rates/status/890609185...
9       https://twitter.com/dog_rates/status/890240255...
10      https://twitter.com/dog_rates/status/890006608...
11      https://twitter.com/dog_rates/status/889880896...
12      https://twitter.com/dog_rates/status/889665388...
13      https://twitter.com/dog_rates/status/889638837...
14      https://twitter.com/dog_rates/status/889531135...
15      https://twitter.com/dog_rates/status/889278841...
16      https://twitter.com/dog_rates/status/888917238...
17      https:

In [172]:
# Quantidqade de urls duplicadas
df_tweets['expanded_urls'].duplicated().value_counts()

False    2219
True      137
Name: expanded_urls, dtype: int64

Foram encontrados 137 url's duplicadas

In [173]:
# Acessando coluna name
df_tweets.name

0        Phineas
1          Tilly
2         Archie
3          Darla
4       Franklin
5           None
6            Jax
7           None
8           Zoey
9         Cassie
10          Koda
11         Bruno
12          None
13           Ted
14        Stuart
15        Oliver
16           Jim
17          Zeke
18       Ralphus
19        Canela
20        Gerald
21       Jeffrey
22          such
23        Canela
24          None
25          None
26          Maya
27        Mingus
28         Derek
29        Roscoe
          ...   
2326       quite
2327           a
2328        None
2329        None
2330        None
2331        None
2332        None
2333          an
2334           a
2335          an
2336        None
2337        None
2338        None
2339        None
2340        None
2341        None
2342        None
2343        None
2344        None
2345         the
2346         the
2347           a
2348           a
2349          an
2350           a
2351        None
2352           a
2353          

In [174]:
df_tweets.name.value_counts

<bound method IndexOpsMixin.value_counts of 0        Phineas
1          Tilly
2         Archie
3          Darla
4       Franklin
5           None
6            Jax
7           None
8           Zoey
9         Cassie
10          Koda
11         Bruno
12          None
13           Ted
14        Stuart
15        Oliver
16           Jim
17          Zeke
18       Ralphus
19        Canela
20        Gerald
21       Jeffrey
22          such
23        Canela
24          None
25          None
26          Maya
27        Mingus
28         Derek
29        Roscoe
          ...   
2326       quite
2327           a
2328        None
2329        None
2330        None
2331        None
2332        None
2333          an
2334           a
2335          an
2336        None
2337        None
2338        None
2339        None
2340        None
2341        None
2342        None
2343        None
2344        None
2345         the
2346         the
2347           a
2348           a
2349          an
2350           a
2351

Acessando a coluna nome é possível identificar dados sem registros e dados incompletos como a, the etc..

In [175]:
df_tweets['in_reply_to_status_id'].value_counts

<bound method IndexOpsMixin.value_counts of 0      NaN
1      NaN
2      NaN
3      NaN
4      NaN
5      NaN
6      NaN
7      NaN
8      NaN
9      NaN
10     NaN
11     NaN
12     NaN
13     NaN
14     NaN
15     NaN
16     NaN
17     NaN
18     NaN
19     NaN
20     NaN
21     NaN
22     NaN
23     NaN
24     NaN
25     NaN
26     NaN
27     NaN
28     NaN
29     NaN
        ..
2326   NaN
2327   NaN
2328   NaN
2329   NaN
2330   NaN
2331   NaN
2332   NaN
2333   NaN
2334   NaN
2335   NaN
2336   NaN
2337   NaN
2338   NaN
2339   NaN
2340   NaN
2341   NaN
2342   NaN
2343   NaN
2344   NaN
2345   NaN
2346   NaN
2347   NaN
2348   NaN
2349   NaN
2350   NaN
2351   NaN
2352   NaN
2353   NaN
2354   NaN
2355   NaN
Name: in_reply_to_status_id, Length: 2356, dtype: float64>

In [176]:
df_tweets[df_tweets['retweeted_status_user_id'].notnull()] 

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19    888202515573088257                    NaN                  NaN   
32    886054160059072513                    NaN                  NaN   
36    885311592912609280                    NaN                  NaN   
68    879130579576475649                    NaN                  NaN   
73    878404777348136964                    NaN                  NaN   
74    878316110768087041                    NaN                  NaN   
78    877611172832227328                    NaN                  NaN   
91    874434818259525634                    NaN                  NaN   
95    873697596434513921                    NaN                  NaN   
97    873337748698140672                    NaN                  NaN   
101   872668790621863937                    NaN                  NaN   
109   871166179821445120                    NaN                  NaN   
118   869988702071779329                    NaN                  NaN   
124   868639477480148993                    NaN                  NaN   
130   867072653475098625                    NaN                  NaN   
132   866816280283807744                    NaN                  NaN   
137   866094527597207552                    NaN                  NaN   
146   863471782782697472                    NaN                  NaN   
155   861769973181624320                    NaN                  NaN   
159   860981674716409858                    NaN                  NaN   
160   860924035999428608                    NaN                  NaN   
165   860177593139703809                    NaN                  NaN   
171   858860390427611136                    NaN                  NaN   
180   857062103051644929                    NaN                  NaN   
182   856602993587888130                    NaN                  NaN   
185   856330835276025856                    NaN                  NaN   
194   855245323840757760                    NaN                  NaN   
195   855138241867124737                    NaN                  NaN   
204   852936405516943360                    NaN                  NaN   
211   851953902622658560                    NaN                  NaN   
...                  ...                    ...                  ...   
784   775096608509886464                    NaN                  NaN   
794   773336787167145985                    NaN                  NaN   
800   772615324260794368                    NaN                  NaN   
811   771171053431250945                    NaN                  NaN   
815   771004394259247104                    NaN                  NaN   
818   770743923962707968                    NaN                  NaN   
822   770093767776997377                    NaN                  NaN   
826   769335591808995329                    NaN                  NaN   
829   768909767477751808                    NaN                  NaN   
833   768554158521745409                    NaN                  NaN   
841   766864461642756096                    NaN                  NaN   
847   766078092750233600                    NaN                  NaN   
860   763167063695355904                    NaN                  NaN   
868   761750502866649088                    NaN                  NaN   
872   761371037149827077                    NaN                  NaN   
885   760153949710192640                    NaN                  NaN   
890   759566828574212096                    NaN                  NaN   
895   759159934323924993                    NaN                  NaN   
908   757729163776290825                    NaN                  NaN   
911   757597904299253760                    NaN                  NaN   
926   754874841593970688                    NaN                  NaN   
937   753298634498793472                    NaN                  NaN   
943   752701944171524096                    NaN                  NaN   
949   752309394570878976                    NaN         

In [177]:
df_tweets[df_tweets['in_reply_to_user_id'].notnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600           8.862664e+17         2.281182e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
113   870726314365509632           8.707262e+17         1.648776e+07   
148   863427515083354112           8.634256e+17         7.759620e+07   
149   863079547188785154           6.671522e+17         4.196984e+09   
179   857214891891077121           8.571567e+17         1.806710e+08   
184   856526610513747968           8.558181e+17         4.196984e+09   
186   856288084350160898           8.562860e+17         2.792810e+08   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
218   850333567704068097           8.503288e+17         2.195506e+07   
228   848213670039564288           8.482121e+17         4.196984e+09   
234   847617282490613760           8.476062e+17         4.196984e+09   
251   844979544864018432           7.590995e+17         4.196984e+09   
274   840698636975636481           8.406983e+17         8.405479e+17   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
346   831926988323639298           8.319030e+17         2.068372e+07   
387   826598799820865537           8.265984e+17         4.196984e+09   
409   823333489516937216           8.233264e+17         1.582854e+09   
427   821153421864615936           8.211526e+17         1.132119e+08   
498   813130366689148928           8.131273e+17         4.196984e+09   
513   811647686436880384           8.116272e+17         4.196984e+09   
565   802265048156610565           7.331095e+17         4.196984e+09   
570   801854953262350336           8.018543e+17         1.185634e+07   
576   800859414831898624           8.008580e+17         2.918590e+08   
611   797165961484890113           7.971238e+17         2.916630e+07   
...                  ...                    ...                  ...   
1479  693582294167244802           6.935722e+17         1.198989e+09   
1497  692423280028966913           6.924173e+17         4.196984e+09   
1501  692142790915014657           6.920419e+17         4.196984e+09   
1523  690607260360429569           6.903413e+17         4.670367e+08   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1605  685681090388975616           6.855479e+17         4.196984e+09   
1618  684969860808454144           6.849598e+17         4.196984e+09   
1630  684538444857667585           6.844811e+17         4.196984e+09   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1689  681340665377193984           6.813394e+17         4.196984e+09   
1774  678023323247357953           6.780211e+17         4.196984e+09   
1819  676590572941893632           6.765883e+17         4.196984e+09   
1842  675870721063669760           6.757073e+17         4.196984e+09   
1844  675849018447167488           6.758457e+17         4.196984e+09   
1852  675707330206547968           6.754971e+17         4.196984e+09   
1866  675349384339542016           6.749998e+17         4.196984e+09   
1882  674999807681908736           6.747934e+17         4.196984e+09   
1885  674793399141146624           6.717299e+17         4.196984e+09   
1892  674754018082705410           6.747522e+17         4.196984e+09   
1895  674742531037511680           6.747400e+17         4.196984e+09   
1905  674606911342424069           6.744689e+17         4.196984e+09   
1914  674330906434379776           6.658147e+17         1.637468e+07   
1940  673716320723169284           6.737159e+17         

In [178]:
df_retweets = df_tweets.tweet_id[ (df_tweets.in_reply_to_user_id.notnull()) | (df_tweets.retweeted_status_user_id.notnull()) ]

In [179]:
df_retweets.count()

259

Existem 259 registros que são retweets.

In [180]:
df_tweets.rating_numerator.describe()

count    2356.000000
mean       13.126486
std        45.876648
min         0.000000
25%        10.000000
50%        11.000000
75%        12.000000
max      1776.000000
Name: rating_numerator, dtype: float64

In [181]:
df_tweets.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [182]:
df_tweets.rating_denominator.describe()

count    2356.000000
mean       10.455433
std         6.745237
min         0.000000
25%        10.000000
50%        10.000000
75%        10.000000
max       170.000000
Name: rating_denominator, dtype: float64

In [183]:
df_tweets.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

A distruibuiçao entre os denominador e numerados são bem variáveis, por exemplo o maior denominador é 170 enquanto o maior numerador é 1776

In [184]:
df_tweets['doggo'].value_counts()

None     2259
doggo      97
Name: doggo, dtype: int64

In [185]:
df_tweets['floofer'].value_counts()

None       2346
floofer      10
Name: floofer, dtype: int64

In [186]:
df_tweets['pupper'].value_counts()

None      2099
pupper     257
Name: pupper, dtype: int64

In [187]:
df_tweets['puppo'].value_counts()

None     2326
puppo      30
Name: puppo, dtype: int64

Temos vários cachorros que não estão dentro das quatro categorias, doggo, floofer, pupper e puppo.

# 2.1.2 - Second Datasource - df_dog_image_predictions - Dog predictions

In [188]:
df_dog_image_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [189]:
df_dog_image_predictions.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [190]:
df_dog_image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [191]:
df_dog_image_predictions.p1_dog.value_counts()

True     1532
False     543
Name: p1_dog, dtype: int64

In [192]:
df_dog_image_predictions.p2_dog.value_counts()

True     1553
False     522
Name: p2_dog, dtype: int64

In [193]:
df_dog_image_predictions.p3_dog.value_counts()

True     1499
False     576
Name: p3_dog, dtype: int64

In [194]:
df_predicition_of_not_dogs = df_dog_image_predictions[ (df_dog_image_predictions.p1_dog == False) & 
                                        (df_dog_image_predictions.p2_dog == False) &
                                        (df_dog_image_predictions.p3_dog == False)
                                      ]

In [195]:
df_predicition_of_not_dogs.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 324 entries, 6 to 2074
Data columns (total 12 columns):
tweet_id    324 non-null int64
jpg_url     324 non-null object
img_num     324 non-null int64
p1          324 non-null object
p1_conf     324 non-null float64
p1_dog      324 non-null bool
p2          324 non-null object
p2_conf     324 non-null float64
p2_dog      324 non-null bool
p3          324 non-null object
p3_conf     324 non-null float64
p3_dog      324 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 26.3+ KB


Segundo o arquivo de predição temos 324 imagens que não foram reconhecidas como contendo cachorros.

In [196]:
# Analisando algumas imagens aleatórias para confirmar que estas imagens não são de cachorros.
df_predicition_of_not_dogs.sample(10)

tweet_id                                          jpg_url  \
455   674767892831932416  https://pbs.twimg.com/media/CV1B0WkWwAEBKVx.jpg   
1851  840370681858686976  https://pbs.twimg.com/media/C6mYrK0UwAANhep.jpg   
226   670361874861563904  https://pbs.twimg.com/media/CU2akCQWsAIbaOV.jpg   
303   671511350426865664  https://pbs.twimg.com/media/CVGwAh-W4AAIHJz.jpg   
56    667065535570550784  https://pbs.twimg.com/media/CUHkkJpXIAA2w3n.jpg   
1606  800459316964663297  https://pbs.twimg.com/media/CxvNfrhWQAA2hKM.jpg   
664   682697186228989953  https://pbs.twimg.com/media/CXltdtaWYAIuX_V.jpg   
1148  731156023742988288  https://pbs.twimg.com/media/CiWWhVNUYAAab_r.jpg   
1240  746906459439529985  https://pbs.twimg.com/media/Cl2LdofXEAATl7x.jpg   
1953  863907417377173506  https://pbs.twimg.com/media/C_03NPeUQAAgrMl.jpg   

      img_num              p1   p1_conf  p1_dog                  p2   p2_conf  \
455         1  shower_curtain  0.238855   False              sarong  0.092414   
1851        1          teapot  0.981819   False                 cup  0.014026   
226         1        platypus  0.974075   False  spotted_salamander  0.011068   
303         1     hermit_crab  0.625409   False                tick  0.127333   
56          1   jigsaw_puzzle  0.560001   False             doormat  0.103259   
1606        1           teddy  0.311928   False            ice_bear  0.184657   
664         1      bald_eagle  0.097232   False               torch  0.096621   
1148        1        lakeside  0.501767   False          breakwater  0.051351   
1240        1   traffic_light  0.470708   False            fountain  0.199776   
1953        1          marmot  0.358828   False             meerkat  0.174703   

      p2_dog                  p3   p3_conf  p3_dog  
455    False              kimono  0.056412   False  
1851   False           coffeepot  0.002421   False  
226    False               bison  0.003897   False  
303    False               snail  0.097916   False  
56     False        space_heater  0.042568   False  
1606   False  Christmas_stocking  0.173229   False  
664    False               cliff  0.090385   False  
1148   False        king_penguin  0.049444   False  
1240   False       space_shuttle  0.064807   False  
1953   False              weasel  0.123485   False

Existem imagens que possuem cachorros porém o modelo de predição não os reconheceu.

In [197]:
df_dog_image_predictions['p1'].value_counts()

golden_retriever             150
Labrador_retriever           100
Pembroke                      89
Chihuahua                     83
pug                           57
chow                          44
Samoyed                       43
toy_poodle                    39
Pomeranian                    38
cocker_spaniel                30
malamute                      30
French_bulldog                26
Chesapeake_Bay_retriever      23
miniature_pinscher            23
seat_belt                     22
Siberian_husky                20
Staffordshire_bullterrier     20
German_shepherd               20
Cardigan                      19
web_site                      19
Shetland_sheepdog             18
teddy                         18
beagle                        18
Eskimo_dog                    18
Maltese_dog                   18
Lakeland_terrier              17
Shih-Tzu                      17
Rottweiler                    17
Italian_greyhound             16
kuvasz                        16
          

In [221]:
df_dog_image_predictions['p2'].value_counts()

Labrador_retriever                104
golden_retriever                   92
Cardigan                           73
Chihuahua                          44
Pomeranian                         42
Chesapeake_Bay_retriever           41
French_bulldog                     41
toy_poodle                         37
cocker_spaniel                     34
miniature_poodle                   33
Siberian_husky                     33
beagle                             28
Eskimo_dog                         27
Pembroke                           27
collie                             27
kuvasz                             26
Italian_greyhound                  22
American_Staffordshire_terrier     21
Pekinese                           21
Samoyed                            20
malinois                           20
toy_terrier                        20
miniature_pinscher                 20
chow                               20
Boston_bull                        19
Norwegian_elkhound                 19
Staffordshir

In [222]:
df_dog_image_predictions['p3'].value_counts()

Labrador_retriever                79
Chihuahua                         58
golden_retriever                  48
Eskimo_dog                        38
kelpie                            35
kuvasz                            34
Staffordshire_bullterrier         32
chow                              32
beagle                            31
cocker_spaniel                    31
Pekinese                          29
toy_poodle                        29
Pomeranian                        29
Pembroke                          27
Great_Pyrenees                    27
Chesapeake_Bay_retriever          27
French_bulldog                    26
malamute                          26
American_Staffordshire_terrier    24
pug                               23
Cardigan                          23
basenji                           21
toy_terrier                       20
bull_mastiff                      20
Siberian_husky                    19
Shetland_sheepdog                 17
Boston_bull                       17
L

In [198]:
df_dog_image_predictions.nunique()

tweet_id    2075
jpg_url     2009
img_num        4
p1           378
p1_conf     2006
p1_dog         2
p2           405
p2_conf     2004
p2_dog         2
p3           408
p3_conf     2006
p3_dog         2
dtype: int64

Existem imagens jpg_url duplicadas.

In [199]:
df_dog_image_predictions.shape

(2075, 12)

In [200]:
df_dog_image_jpg_url_duplicated = df_dog_image_predictions[df_dog_image_predictions.jpg_url.duplicated(keep=False)]

In [201]:
df_dog_image_jpg_url_duplicated.shape

(132, 12)

Temos 132 imagens duplicadas no dataFrame de predição de imagens

# 2.1.3 - Third Datasource - df_tweets_api - Data gathered from Twitter API

In [203]:
df_tweets_api.head()

favorite_count  retweet_count            tweet_id
0           37589           8198  892420643555336192
1           32309           6060  892177421306343424
2           24348           4009  891815181378084864
3           40933           8357  891689557279858688
4           39126           9049  891327558926688256

In [204]:
df_tweets_api.info().info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2337 entries, 0 to 2336
Data columns (total 3 columns):
favorite_count    2337 non-null int64
retweet_count     2337 non-null int64
tweet_id          2337 non-null int64
dtypes: int64(3)
memory usage: 54.9 KB


In [205]:
df_tweets_api.describe()

favorite_count  retweet_count      tweet_id
count     2337.000000    2337.000000  2.337000e+03
mean      7855.413778    2882.041078  7.420692e+17
std      12175.902322    4868.749782  6.824243e+16
min          0.000000       1.000000  6.660209e+17
25%       1366.000000     580.000000  6.783345e+17
50%       3420.000000    1346.000000  7.185406e+17
75%       9626.000000    3358.000000  7.986825e+17
max     162407.000000   82576.000000  8.924206e+17

### Quality

1. Na listagem existem tweets e retweets, será necessário remover os retweets.
2. A coluna name possui valores incorretos.
3. As colunas que representam o numerador e denominador possuem valores que distorcem a análise.
4. A coluna timestamp está como objeto ao invez de um datetime.
5. O dados da coluna source estão como HTML
6. A coluna text possui links, será necessário separar em um nova coluna.
7. Existem imagens que não são de caes.
8. Existe dados duplicados para a coluna jpg_url
9. Existem dados duplicados em todos os dataframes.
10. Tipos de dados inconsistentes.
11. Inconsistencia no nome das raças possuindo underscore e hora com a primeira letra em maisculo hora em minusculo.


### Tidiness

1. As colunas para o dog stage deveriam ser apenas uma coluna.
2. Existem dadods irrelevantes ou que não serão utilizados na análise
3. Todos os 3 dataframes possuem tamanhos diferentes.
4. Os dados estão dividos em 3 dataframes

## 3. Clean

## 4. Explore

## 5. Conclusion